### PR TITLE
Next Version

### PR DESCRIPTION
**Pull request by sendra**
Moved MessageType to a new library and refactored code to use it.
Created a new Parent contract CrossChainAdapter so as to abstract communication there.

Added message types and usage for vote results and payload execution

### CODE REVIEW

File: src/contracts/libraries/BridgingHelper.sol [hunk 1]
```diff
@@ -15,4 +17,83 @@ library BridgingHelper {
...
+
+  function decodeMessageWithType(
+    bytes memory messageWithType
+  ) internal pure returns (BridgingHelper.MessageType, bytes memory) {
+    return abi.decode(messageWithType, (BridgingHelper.MessageType, bytes));
+  }
+
+  function decodeStartProposalVoteMessage(
+    bytes memory message
+  ) internal pure returns (uint256, bytes32, uint24) {
```
Comment (kyzia551) [inline src/contracts/libraries/BridgingHelper.sol:29]: lucking documentation, or maybe make sense to have it as struct, not sure

File: src/contracts/libraries/BridgingHelper.sol [hunk 2]
```diff
@@ -15,4 +17,83 @@ library BridgingHelper {
...
+      abi.decode(
+        message,
+        (uint40, PayloadsControllerUtils.AccessControl, uint40)
+      );
+  }
+
+  function decodeVoteResultMessage(
+    bytes memory message
+  ) internal pure returns (uint256, uint128, uint128) {
+    return abi.decode(message, (uint256, uint128, uint128));
```
Comment (kyzia551) [inline src/contracts/libraries/BridgingHelper.sol:50]: lucking documentation, or maybe make sense to have it as struct, not sure

File: src/contracts/libraries/BridgingHelper.sol [hunk 3]
```diff
@@ -15,4 +17,83 @@ library BridgingHelper {
...
+  ) internal pure returns (uint256, bytes32, uint24) {
+    return abi.decode(message, (uint256, bytes32, uint24));
+  }
+
+  function decodePayloadExecutionMessage(
+    bytes memory message
+  )
+    internal
+    pure
+    returns (uint40, PayloadsControllerUtils.AccessControl, uint40)
```
Comment (kyzia551) [inline src/contracts/libraries/BridgingHelper.sol:38]: lucking documentation, or maybe make sense to have it as struct, not sure

File: src/contracts/libraries/BridgingHelper.sol [hunk 4]
```diff
@@ -15,4 +17,83 @@ library BridgingHelper {
...
+      );
+  }
+
+  function decodeVoteResultMessage(
+    bytes memory message
+  ) internal pure returns (uint256, uint128, uint128) {
+    return abi.decode(message, (uint256, uint128, uint128));
+  }
+
+  function encodePayloadMessageForExecution(
```
Comment (kyzia551) [inline src/contracts/libraries/BridgingHelper.sol:53]: better to name `encodePayloadExecutionMessage`, as you did in corresponding decode

**Review by kyzia551 — CHANGES_REQUESTED** [review]

File: src/contracts/voting/VotingMachine.sol [hunk 5]
```diff
@@ -184,4 +116,60 @@ contract VotingMachine is VotingMachineWithProofs, IVotingMachine {
...
+  ) internal override {
+    bytes memory empty;
+    if (messageType == BridgingHelper.MessageType.Proposal_Vote) {
+      try this.decodeStartProposalVoteMessage(message) returns (
+        uint256 proposalId,
+        bytes32 blockHash,
+        uint24 votingDuration
+      ) {
+        _createBridgedProposalVote(proposalId, blockHash, votingDuration);
+        emit MessageReceived(
```
Comment (kyzia551) [inline src/contracts/voting/VotingMachine.sol:148]: maybe we can abstract this part as well, checking that msg.sender is address(this)

File: src/contracts/MessageWithTypeReceiver.sol [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        work with the bridged message
+ */
+abstract contract MessageWithTypeReceiver is IMessageWithTypeReceiver {
+  /// @inheritdoc IBaseReceiverPortal
+  function receiveCrossChainMessage(
+    address originSender,
+    uint256 originChainId,
+    bytes memory messageWithType
+  ) external {
+    _checkOrigin(msg.sender, originSender, originChainId);
```
[the remark this answers is not in the document]
Comment (sendra) [reply src/contracts/MessageWithTypeReceiver.sol]: Updated with the check returning bool and moving the require up

File: src/contracts/voting/VotingMachine.sol (hunk 5 above)
Comment (sendra) [reply src/contracts/voting/VotingMachine.sol:146]: Did not find a way on abstracting, as condition is specific on the inheriting contract

**Comment by sendra** [comment]
Closing as this will not be implemented for now

**Pull request closed without merging**